org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/markers/ApplyQuickFixMarkerResolver.java

```diff
@@ -99,17 +133,25 @@ private static void applyIn(IDocument document, MarkerQuickFix fix) {
     }
   }
 
-  private static void apply(IDocumentExtension4 document, MarkerTextEdit textEdit) {
+  private static void apply(ITextEditor textEditor, IDocumentExtension4 document, MarkerTextEdit textEdit) {
     IMarker editMarker = textEdit.getMarker();
     try {
-      ITextEditor textEditor = LocationsUtils.findOpenEditorFor(editMarker);
-      if (textEditor == null) {
-        // TODO handle closed editors
-        return;
-      }
-      Position markerPosition = LocationsUtils.getMarkerPosition(editMarker, textEditor);
-      if (markerPosition != null) {
-        document.replace(markerPosition.getOffset(), markerPosition.getLength(), textEdit.getNewText(), document.getModificationStamp() + 1);
+      if (editMarker.exists()) {
+        int start = MarkerUtilities.getCharStart(editMarker);
+        int end = MarkerUtilities.getCharEnd(editMarker);
+        // look up the current range of the marker when the document has been edited
+        Position markerPosition = LocationsUtils.getMarkerPosition(editMarker, textEditor);
+        if (markerPosition != null) {
+          if (markerPosition.isDeleted()) {
+            // do nothing if position has been deleted
+            return;
+          } else {
+            // use position instead of marker values
+            start = markerPosition.getOffset();
+            end = markerPosition.getOffset() + markerPosition.getLength();
+          }
+        }
+        document.replace(start, end - start, textEdit.getNewText(), document.getModificationStamp() + 1);
```

Comment:
We should maybe have some logs in some cases here. Else it means the application of the fix could have no effect at all (no change in the document and no log)

Reply:
Now that I think about it, I think we could probably move/replay some of those checks earlier when the resolvers are contributed. The `IMarkerResolutionGenerator2` is called just when the fixes are about to be displayed, so it could be a good time to check those cases. This would let us not propose a fix instead of proposing it knowing it won't have any effect